Upload windows docker images docs (#249)
game-ci/documentation: docs/docker/02-windows-docker-images.md

# Windows docker images

Windows images are used for some features available only in windows. The main reason is usually building an IL2CPP version of the application.

> To build applications effectively in terms of build time, the very base image is a .NET framework 4.8. It's cached on GitHub actions runners and we truly need .NET. The base Windows image can be found here [`windows/base/Dockerfile`](https://github.com/game-ci/docker/blob/main/images/windows/base/Dockerfile)

Reviewer: ```suggestion
To build applications effectively in terms of build time, the very base image is a .NET framework 4.8. It's cached on GitHub actions runners and we truly need .NET. The base Windows image can be found in [https://github.com/game-ci/docker](https://github.com/game-ci/docker/blob/main/images/windows/base/Dockerfile)
```

> 
> Images for newly released Unity editor versions are added almost immediately to [unityci](https://hub.docker.com/u/unityci). This process is automated, please allow a few hours.

Reviewer: ```suggestion
```

> Windows images have different editor template. It's `windows-{UNITY_EDITOR_VERSION}-{MODULE_NAME}-{GAMECI_RELEASE_VERSION}`. E.g. `windows-2021.2.15f1-windows-il2cpp-0.17.2`

Reviewer: This is actually not true. Ubuntu is using the same scheme.

In addition `ubuntu` is the default build platform for most target platforms, so we also have tags without `ubuntu` for those. In the future we might do the same for windows images, in case of `windows-il2cpp` and some others.

Recommend  leaving this sentence away entirely and stick to a single source of truth (which is also represented  on the https://game.ci/docs/docker/versions page)

## Features

- Git
- Choco package manager

## Limitations

> ### GameCI can't redistribute Visual Studio in their docker images

Reviewer: ```suggestion
### Microsoft does not allow redistribution of Visual Studio build tools
```

Microsoft doesn't permit publishing images with VS Build tools [`discussion`](https://github.com/microsoft/Windows-Containers/issues/102#issuecomment-827170844) so aren't we allowed to do that. Building windows IL2CPP project in pure editor images leads to exceptions claiming that "C++ code builder" nor "Windows 10 SDK" are installed. There are two options to avoid this obstacle

1. Installs vctools, e.g. `choco install visualstudio2017-workload-vctools --no-progress -y` and push it to a private docker registry

> 2. Use Microsoft Visual Studio with VC++ and Windows 10 SDK from the host system

Reviewer: ```suggestion
2. Inject/mount Microsoft Visual Studio with VC++ and Windows 10 SDK from the host system into the container during runtime.
```

> 

Reviewer: ```suggestion

### Injecting VC++ into the image

```

```toml
...
volumes = [
  "C:\\Program Files (x86)\\Windows Kits:C:\\Program Files (x86)\\Windows Kits",
  "C:\\Program Files (x86)\\Microsoft Visual Studio:C:\\Program Files (x86)\\Microsoft Visual Studio",
  "C:\\Program Files\\Microsoft Visual Studio:C:\\Program Files\\Microsoft Visual Studio",
  "C:\\ProgramData\\Microsoft\\VisualStudio:C:\\ProgramData\\Microsoft\\VisualStudio"
]
...
```

### Licenses are not valid between containers runs

This behaviour is different to Ubuntu containers where it's possible to use a single licence for even a few different runners.

In Windows, it's necessary to acquire a license every time and return it after a building process. License files for every run are identical apart from the last four symbols of the machine hash code. This restriction has not been solved yet

The simplest way is to write a couple of scripts (copied from [`Unity3d docs`](https://docs.unity3d.com/Manual/ManagingYourUnityLicense.html)) and wrap it to the try-finally block

**Note**: Path to Unity3d is `C:\Program Files\Unity\Hub\Editor\{UNITY_EDITOR_VERSION}\Editor\Unity.exe`

```pwsh
try {
  build.ps1
} finally {
  return-license.ps1
}
```

> ## Gitlab-specific problems

Reviewer: ```suggestion
## Troubleshooting: GitLab
```
